core/qbft/qbft.go

```diff
@@ -202,6 +203,29 @@ func Run[I any, V comparable](ctx context.Context, d Definition[I, V], t Transpo
 		dedupIn = dedup
 	}
 
+	isDuplicatedRule := func(rule uponRule) bool {
+		if rule == uponJustifiedDecided {
+			return false
+		}
+		if dedupRules[rule] {
+			return true
+		}
+
+		// first time for this rule
+		dedupRules[rule] = true
+
+		return false
+	}
+
+	resetDedupRules := func() {
+		dedupRules = make(map[uponRule]bool)
+	}
+
+	changeRound := func(newRound int64) {
+		resetDedupRules()
```

Comment:
can probably inline `resetDedupRules` since it is just a single line function, and I would also move `trimBuffer` here:
```
// changeRound updates the round and clears state of the previous round
changeRound := func(newRound int64) {
    round = newRound
    dedupRules = make(map[uponRule]bool)
    trimBuffer()
}
```

Reply:
Done